File: CC_Visual Basic/VisualBasic/BeginVBE/BeginVBE/BVBE Projects/Decode/Form1.vb
Public Class Form1
  Dim GameOver As Boolean
  Dim MyGuess As String
  Dim ComputerCode As String
  Dim ComputerNumbers(4) As String
  Dim MyRandom As New Random
  Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    lblMessage.Text = "Click New Game"
    btnNew.Focus()
  End Sub
  Private Sub btnNew_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnNew.Click
    Dim NArray(10) As String
    Dim T As Integer
    Dim I As Integer, J As Integer
    'Start new game
    btnStop.Enabled = True
    GameOver = False
    lblMessage.Text = ""
    txtGuesses.Text = ""
    txtGuess.Text = ""
    btnNew.Enabled = False
    'Choose code using modified version of card shuffling routine
    'Order all digits initially
    ComputerCode = ""
    For I = 1 To 10 : NArray(I) = I - 1 : Next I
    'J is number of integers remaining
    For J = 10 To 7 Step -1
      I = MyRandom.Next(J) + 1
      ComputerNumbers(11 - J) = NArray(I)
      ComputerCode = ComputerCode & NArray(I)
      T = NArray(J)
      NArray(J) = NArray(I)
      NArray(I) = T
    Next J
    lblMessage.Text = "I have a 4 digit code." & ControlChars.CrLf & "Try to guess it."
    grpGuess.Visible = True
    txtGuess.Focus()
  End Sub
  Private Sub btnStop_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnStop.Click
    'Stop current game
    grpGuess.Visible = False
    btnNew.Enabled = True
    btnStop.Enabled = False
    If Not (GameOver) Then
      lblMessage.Text = "Game Stopped" & ControlChars.CrLf & "My code was - " & ComputerCode
    End If
    btnNew.Focus()
  End Sub
  Private Sub txtGuess_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles txtGuess.KeyPress
    'Allow numbers only
    If e.KeyChar = ControlChars.Cr Then
      btnCheck.PerformClick()
    ElseIf (e.KeyChar >= "0" And e.KeyChar <= "9") Or e.KeyChar = ControlChars.Back Then
      e.Handled = False
    Else
      e.Handled = True
    End If
  End Sub
  Private Sub btnCheck_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCheck.Click
    Dim W As String
    Dim WNumbers(4) As String
    Dim I As Integer, J As Integer, K1 As Integer, K2 As Integer
    Dim Distinct As Boolean
    'Check your guess
    W = txtGuess.Text
    'Check length validity
    If Len(W) <> 4 Then
      lblMessage.Text = "Guess must have 4 numbers ..." & ControlChars.CrLf & "Try again!"
      txtGuess.Focus()
      Exit Sub
    Else
      'Get numbers and make sure they are distinct
      Distinct = True
      For I = 1 To 4
        WNumbers(I) = Mid(W, I, 1)
        If I <> 1 Then
          For J = I - 1 To 1 Step -1
            If WNumbers(I) = WNumbers(J) Then
              Distinct = False
            End If
          Next J
        End If
      Next I
      If Not (Distinct) Then
        lblMessage.Text = "Numbers must all be different ..." & ControlChars.CrLf & "Try again!"
        txtGuess.Focus()
        Exit Sub
      End If
      If W = ComputerCode Then
        lblMessage.Text = "Congratulations!" & ControlChars.CrLf & "That's my code - " & ComputerCode
        GameOver = True
        btnStop.PerformClick()
        Exit Sub
      Else
        'Compute score
        K1 = 0
        K2 = 0
        For I = 1 To 4
          For J = 1 To 4
            If WNumbers(J) = ComputerNumbers(I) Then K1 = K1 + 1
          Next J
          If WNumbers(I) = ComputerNumbers(I) Then K2 = K2 + 1
        Next I
        lblMessage.Text = "Your guess - " & W & ControlChars.CrLf & Format(K1, "0") & " digit(s) correct" & ControlChars.CrLf & Format(K2, "0") & " digit(s) in proper place"
        txtGuesses.Text = W & "   " & Format(K1, "0") & "/" & Format(K2, "0") & ControlChars.CrLf & txtGuesses.Text
        txtGuess.Text = ""
        txtGuess.Focus()
      End If
    End If
  End Sub


End Class

File: CC_C#/BVCS Projects/Decode/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;

#endregion

namespace Decode
{
    partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        bool gameOver;
        string computerCode;
        string[] computerNumbers = new string[4];
        Random myRandom = new Random();

        private void Form1_Load(object sender, EventArgs e)
        {
            lblMessage.Text = "Click New Game";
            btnNew.Focus();
        }

        private void btnNew_Click(object sender, EventArgs e)
        {
            string[] nArray = new string[10];
            int i, j;
            string t;
            // Start new game
            btnStop.Enabled = true;
            gameOver = false;
            lblMessage.Text = "";
            txtGuesses.Text = "";
            txtGuess.Text = "";
            btnNew.Enabled = false;
            // Choose code using modified version of card shuffling routine
            // Order all digits initially
            computerCode = "";
            for (i = 0; i < 10; i++)
            {
                nArray[i] = Convert.ToString(i);
            }
            // J is number of integers remaining
            for (j = 9; j >=6; j--)
            {
                i = myRandom.Next(j);
                computerNumbers[9 - j] = nArray[i];
                computerCode = computerCode + nArray[i];
                t = nArray[j];
                nArray[j] = nArray[i];
                nArray[i] = t;
            }
            lblMessage.Text = "I have a 4 digit code.\r\nTry to guess it.";
            grpGuess.Visible = true;
            txtGuess.Focus();
        }

        private void btnStop_Click(object sender, EventArgs e)
        {
            // Stop current game
            grpGuess.Visible = false;
            btnNew.Enabled = true;
            btnStop.Enabled = false;
            if (!gameOver)
            {
                lblMessage.Text = "Game Stopped\r\nMy code was - " + computerCode;
            }
            btnNew.Focus();
        }

        private void txtGuess_KeyPress(object sender, KeyPressEventArgs e)
        {
            // Allow numbers only
            if ((int) e.KeyChar == 13)
            {
                btnCheck.PerformClick();
            }
            else if ((e.KeyChar >= '0' && e.KeyChar <= '9') || (int) e.KeyChar == 8)
            {
                e.Handled = false;
            }
            else
            {
                e.Handled = true;
            }
        }

        private void btnCheck_Click(object sender, EventArgs e)
        {
            string w;
            string[] wNumbers = new string[4];
            int i, j, k1, k2;
            bool distinct;
            // Check your guess
            w = txtGuess.Text;
            // Check length validity
            if (w.Length != 4)
            {
                lblMessage.Text = "Guess must have 4 numbers ...\r\nTry again!";
                txtGuess.Focus();
                return;
            }
            else
            {
                // Get numbers and make sure they are distinct
                distinct = true;
                for (i = 0; i < 4; i++)
                {
                    wNumbers[i] = w.Substring(i, 1);
                    if (i != 0)
                    {
                        for (j = i - 1; j >= 0 ; j--)
                        {
                            if (wNumbers[i] == wNumbers[j])
                            {
                                distinct = false;
                            }
                        }
                    }
                }
                if (!distinct)
                {
                    lblMessage.Text = "Numbers must all be different ...\r\nTry again!";
                    txtGuess.Focus();
                    return;
                }
                if (w == computerCode)
                {
                    lblMessage.Text = "Congratulations!\r\nThat's my code - " + computerCode;
                    gameOver = true;
                    btnStop.PerformClick();
                    return;
                }
                else
                {
                    // Compute score
                    k1 = 0;
                    k2 = 0;
                    for (i = 0; i < 4; i++)
                    {
                        for (j = 0; j < 4; j++)
                        {
                            if (wNumbers[j] == computerNumbers[i])
                                k1++;
                        }
                        if (wNumbers[i] == computerNumbers[i])
                            k2++;
                    }
                    lblMessage.Text = "Your guess - " + w + "\r\n" + Convert.ToString(k1) + " digit(s) correct\r\n" + Convert.ToString(k2) + " digit(s) in proper place";
                    txtGuesses.Text = w + "   " + Convert.ToString(k1) + "/" + Convert.ToString(k2) + "\r\n" + txtGuesses.Text;
                    txtGuess.Text = "";
                    txtGuess.Focus();
                }
            }
        }
    }
}